src/keymap/keymap.rs

```diff
@@ -21,10 +22,17 @@ use crate::state::State;
 pub struct Keymap(*mut xkb_keymap);
 
 impl Keymap {
-	pub fn from_str(s: &str) -> Option<Self> {
+	pub fn from_str(s: &CStr) -> Option<Self> {
 		unsafe{
 			let context = xkb_context_new(0);
-			let ptr = xkb_keymap_new_from_string(context, s.as_bytes().as_ptr() as *const _ as *const std::os::raw::c_char, xkb_keymap_format::XKB_KEYMAP_FORMAT_TEXT_v1, 0);
+			if context == std::ptr::null_mut(){
+				return None;
+			}
+
+			let ptr = xkb_keymap_new_from_string(context, s.to_bytes().as_ptr() as *const _ as *const std::os::raw::c_char, xkb_keymap_format::XKB_KEYMAP_FORMAT_TEXT_v1, 0);
```

Comment:
Next question is if `xkb_keymap_new_from_string` bumps the refcount for the context or not, this is why I got stuck with adding this from an API standpoint, usually the context is taken from X11 and not created from thin air.

What does the documentation say about this?

Reply:
We can put it in the argument if you want. I only need this for Wayland so I have no clue about the usage in X11.

Reply:
We then would have a constructor like this:
"fn from_str(s: &CStr, context: &Context) -> Option<Self>"

Reply:
But then the only way to get a `Context` at the moment is through the `x11` stuff, basically it's an API design issue.

Do you have some sample C code to load a keymap for Wayland?

Reply:
But you can create a new context with https://docs.rs/xkb/0.2.1/xkb/struct.Context.html#method.new
I can only link you to the weston code otherwise Ive written it with some FFI rust code.
https://github.com/wayland-project/weston/blob/master/clients/window.c#L2996-L3091

Reply:
In Weston they reuse the input XKB context instead of creating a new temporary one.

How does Weston create the context? Do they keep it around all the time? Is it using `xkb_context_new`?

I guess `Keymap::from_str(&Context, &str)` would be fine by me.

Reply:
https://github.com/wayland-project/weston/blob/master/clients/window.c#L6247

Reply:
Really a &str not a &CStr? As I am unsure if this is an UTF-8 String.

Reply:
Internally it needs to be converted to a `CString` before getting passed to `xkb_keymap_from_string`, but externally you don't want people to care about creating a `CString`.

Also I just checked the source of xkbcommon and internally `xkb_keymap_new` refs the context, so I'd go with two versions:

`Keymap::from_str(&Context, &str) -> Result<Keymap>` and `impl FromStr for Keymap` (this would allow using `str::parse`).

The `FromStr` implementation creates a new context and uses `Keymap::from_str` to create the `Keymap`.

Reply:
How did you find out that it creates a CString internally?

Reply:
Adjusted it just now.